[BUG] Searxng: URI illegal character

<!-- Please provide as many details as possible, this will help us to deliver a fix as soon as possible. Thank you! -->

## Describe the bug
<!-- A clear and concise description of what the bug is. -->

`langchain4j-community-web-search-engine-searxng` can't handle illegal character in URI properly.

## Log and Stack trace
<!-- Please provide a log and a stack trace (with exception), if applicable. -->

```
java.lang.IllegalArgumentException: Illegal character in fragment at index 119: https://www.linkedin.com/pulse/introduction-langchain4j-supercharging-java-llms-ibrahim-jimoh-kyj4f#:~:text=LangChain4j is a groundbreaking Java,by Python and JavaScript libraries.
	at java.base/java.net.URI.create(URI.java:906)
	at dev.langchain4j.community.web.search.searxng.SearXNGWebSearchEngine.toWebSearchOrganicResult(SearXNGWebSearchEngine.java:59)
	at dev.langchain4j.community.web.search.searxng.SearXNGWebSearchEngine.lambda$search$1(SearXNGWebSearchEngine.java:79)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1602)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at dev.langchain4j.community.web.search.searxng.SearXNGWebSearchEngine.search(SearXNGWebSearchEngine.java:79)
	at dev.langchain4j.web.search.WebSearchEngine.search(WebSearchEngine.java:15)
	at dev.langchain4j.web.search.WebSearchEngineIT.should_search(WebSearchEngineIT.java:22)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.net.URISyntaxException: Illegal character in fragment at index 119: https://www.linkedin.com/pulse/introduction-langchain4j-supercharging-java-llms-ibrahim-jimoh-kyj4f#:~:text=LangChain4j is a groundbreaking Java,by Python and JavaScript libraries.
	at java.base/java.net.URI$Parser.fail(URI.java:2976)
	at java.base/java.net.URI$Parser.checkChars(URI.java:3147)
	at java.base/java.net.URI$Parser.parse(URI.java:3191)
	at java.base/java.net.URI.<init>(URI.java:623)
	at java.base/java.net.URI.create(URI.java:904)
	... 18 more
```

## Minimal Example to Reproduce
<!-- Please provide a relevant code snippets to reproduce this bug. -->

```java
String url = "https://www.linkedin.com/pulse/introduction-langchain4j-supercharging-java-llms-ibrahim-jimoh-kyj4f#:~:text=LangChain4j is a groundbreaking Java,by Python and JavaScript libraries."

URI.create(url)
```

## Expected behavior
<!-- A clear and concise description of what you expected to happen. -->

Handle illegal character properly. (Remove or encode them?)

## Please complete the following information:
- LangChain4j Community version: 0.36.2
- LangChain4j Community Component you used (e.g. `langchain4j-community-clickhouse`): `langchain4j-community-web-search-engine-searxng`
- Java version: 17
- Spring Boot version (if applicable): 

## Additional context
<!-- Add any other context about the problem here. -->

/cc @bnayfeh

--- a/web-search-engines/langchain4j-community-web-search-engine-searxng/src/main/java/dev/langchain4j/community/web/search/searxng/SearXNGWebSearchEngine.java
+++ b/web-search-engines/langchain4j-community-web-search-engine-searxng/src/main/java/dev/langchain4j/community/web/search/searxng/SearXNGWebSearchEngine.java
@@ -49,6 +49,13 @@ public class SearXNGWebSearchEngine implements WebSearchEngine {
         return String.join(",", values.toString());
     }
 
+    static URI makeURI(String urlString) {
+        if (urlString == null || urlString.isBlank()) {
+            throw new IllegalArgumentException("urlString can not be null or blank");
+        }
+        return URI.create(urlString.replaceAll("\\s+", "%20"));
+    }
+
     private static Map<String, String> extractMetadata(SearXNGResult result) {
         final Map<String, String> metadata = new HashMap<>();
         metadata.put("engine", result.getEngine());
@@ -60,13 +67,8 @@ public class SearXNGWebSearchEngine implements WebSearchEngine {
     }
 
     private static WebSearchOrganicResult toWebSearchOrganicResult(SearXNGResult result) {
-        // FIXME: temporarily fix URI illegal character, raise a issue to solve it.
-        String url = result.getUrl();
-        int illegalChar = url.indexOf('#');
-        url = illegalChar == -1 ? url : url.substring(0, illegalChar);
-
         return WebSearchOrganicResult.from(
-                result.getTitle(), URI.create(url), result.getContent(), null, extractMetadata(result));
+                result.getTitle(), makeURI(result.getUrl()), result.getContent(), null, extractMetadata(result));
     }
 
     private static boolean hasValue(String value) {

--- a/web-search-engines/langchain4j-community-web-search-engine-searxng/src/test/java/dev/langchain4j/community/web/search/searxng/SearXNGWebSearchEngineTest.java
+++ b/web-search-engines/langchain4j-community-web-search-engine-searxng/src/test/java/dev/langchain4j/community/web/search/searxng/SearXNGWebSearchEngineTest.java
@@ -1,0 +1,55 @@
+package dev.langchain4j.community.web.search.searxng;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import java.net.URI;
+import org.junit.jupiter.api.Test;
+
+class SearXNGWebSearchEngineTest {
+
+    private static void testURI(String uriString) {
+        assertDoesNotThrow(() -> SearXNGWebSearchEngine.makeURI(uriString));
+        final URI uri = SearXNGWebSearchEngine.makeURI(uriString);
+        assertNotNull(uri);
+        if (uriString.matches(".*\\s+.*")) {
+            assertNotEquals(uriString, uri.toString());
+        } else {
+            assertEquals(uriString, uri.toString());
+        }
+    }
+
+    @Test
+    void test_malformed_urls() {
+        assertThrows(
+                IllegalArgumentException.class,
+                () -> URI.create(
+                        "https://www.linkedin.com/pulse/introduction-langchain4j-supercharging-java-llms-ibrahim-jimoh-kyj4f#:~:text=LangChain4j is a groundbreaking Java,by Python and JavaScript libraries."));
+        assertDoesNotThrow(
+                () -> URI.create(
+                        "https://www.linkedin.com/pulse/introduction-langchain4j-supercharging-java-llms-ibrahim-jimoh-kyj4f#:~:text=LangChain4j"));
+        assertDoesNotThrow(() -> URI.create(
+                "https://www.linkedin.com/pulse/introduction-langchain4j-supercharging-java-llms-ibrahim-jimoh-kyj4f#:~:text=LangChain4j is a groundbreaking Java,by Python and JavaScript libraries."
+                        .replaceAll(" ", "%20")));
+        assertDoesNotThrow(() -> URI.create(
+                "https://www.linkedin.com/pulse/introduction-langchain4j-supercharging-java-llms-ibrahim-jimoh-kyj4f#:~:text=LangChain4j is a groundbreaking Java,by Python and JavaScript libraries."
+                        .replaceAll("\\s+", "%20")));
+        assertDoesNotThrow(() -> URI.create(
+                "https://www.linkedin.com/pulse/introduction-langchain4j-supercharging-java-llms-ibrahim-jimoh-kyj4f"));
+    }
+
+    @Test
+    void test_make_uri() {
+        assertThrows(IllegalArgumentException.class, () -> SearXNGWebSearchEngine.makeURI(null));
+        assertThrows(IllegalArgumentException.class, () -> SearXNGWebSearchEngine.makeURI(""));
+        assertThrows(IllegalArgumentException.class, () -> SearXNGWebSearchEngine.makeURI("  \\t"));
+        testURI(
+                "https://www.linkedin.com/pulse/introduction-langchain4j-supercharging-java-llms-ibrahim-jimoh-kyj4f#:~:text=LangChain4j");
+        testURI(
+                "https://www.linkedin.com/pulse/introduction-langchain4j-supercharging-java-llms-ibrahim-jimoh-kyj4f#:~:text=LangChain4j is a groundbreaking Java,by Python and JavaScript libraries.");
+        testURI("https://www.linkedin.com/pulse/introduction-langchain4j-supercharging-java-llms-ibrahim-jimoh-kyj4f");
+    }
+}

